### ISSUE TITLE
asyncio support?

### ISSUE THREAD

**Issue by adaamz** [schireson/pytest-alembic#44]
Hi,
I would like to ask if is it possible to use asynchronous sqlalchemy with pytest-alembic. If yes then how to configure it, because I'm not able to configure it proper way.

This is how I create session in tests:
```python
@pytest.fixture()
async def async_session_empty(config):
    engine = create_async_engine(config.db.url)
    async_sess = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)
    async with async_sess() as session:
        yield session

    await engine.dispose()
```

I tried this
```python
@pytest.fixture()
@pytest.mark.asyncio
async def alembic_engine(async_session_empty):
    return async_session_empty


@pytest.fixture()
def alembic_config(config):
    # use this in env.py
    return {"script_location": "alembic", "db_url": config.db.url}
```
but I only get some pytest warning about unclosed socket
```
ResourceWarning: unclosed <socket.socket fd=22, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.0.6', 57792), raddr=('172.20.0.3', 5432)>
```

Thank you for your response.

**Comment by DanCardin** [comment]
The answer is going to minimally include applying https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html to your `env.py` if you haven't already.

Although in my attempt at writing a test-case for this scenario so far, I am personally seeing:
```
tests::pytest_alembic::test_model_definitions_match_ddl <- . FAILED      [ 25%]
tests::pytest_alembic::test_single_head_revision <- . PASSED             [ 50%]
tests::pytest_alembic::test_up_down_consistency <- . FAILED              [ 75%]
tests::pytest_alembic::test_upgrade <- . FAILED                          [100%]

=================================== FAILURES ===================================
___ [pytest-alembic] tests::pytest_alembic::test_model_definitions_match_ddl ___
Alembic Error:
    Task <Task pending name='Task-3' coro=<run_migrations_online() running at migrations/env.py:27> cb=[_run_until_complete_cb() at ...python3-3.8.8/lib/python3.8/asyncio/base_events.py:184]> got Future <Future pending cb=[Protocol._on_waiter_completed()]> attached to a different loop
```
So it appears to be slightly working.

In any case, the internals of any of the fanciers features of this plugin are currently assuming sync sqlalchemy, so once I get a basic test passing, i'll have to noodle around on whether the rest is supportable

**Comment by DanCardin** [comment]
I'll be pushing a new patch version which fixes some of the bells and whistles and documents the requirements, but essentially https://alembic.sqlalchemy.org/en/latest/cookbook.html#using-asyncio-with-alembic is correct, except that the example appears to be slightly out of date and should instead look like this:

```python
def do_run_migrations(connection):
    context.configure(connection=connection, target_metadata=target_metadata)

    with context.begin_transaction():
        context.run_migrations()


async def run_migrations_online():
    """Run migrations in 'online' mode.

    In this scenario we need to create an Engine
    and associate a connection with the context.

    """
    connectable = AsyncEngine(
        engine_from_config(
            config.get_section(config.config_ini_section),
            prefix="sqlalchemy.",
            poolclass=pool.NullPool,
            future=True,
        )
    )

    async with connectable.connect() as connection:
        await connection.run_sync(do_run_migrations)

    # you also need this line
    await connectable.dispose()


if context.is_offline_mode():
    run_migrations_offline()
else:
    asyncio.run(run_migrations_online())
```

Bayond that, in your example:
```
@pytest.fixture()
@pytest.mark.asyncio
async def alembic_engine(async_session_empty):
    return async_session_empty
```

`async_session_empty` needs to be an engine. So you can either avoid creating the session entirely, or if you use elsewhere, you might want:

```
@pytest.fixture()
async def async_engine(config):
    engine = create_async_engine(config.db.url)
    yield engine
    await engine.dispose()

@pytest.fixture()
async def async_session_empty(async_engine):
    engine = create_async_engine(config.db.url)
    async_sess = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)
    async with async_sess() as session:
        yield session

@pytest.fixture()
@pytest.mark.asyncio
async def alembic_engine(async_engine):
    return async_engine

# or perhaps, this might work
alembic_engine = async_engine
```

Note that we're ultimately hooking into alembic's mechanisms, and there's no way around that (that i'm aware of), so it's going to `COMMIT` the changes per test, and your engine/database setup needs to ensure a clean state for each test

**Comment by DanCardin** [comment]
released extra support in 0.7.0, and updated docs. but my above suggestions should work for the default tests in your current version. lmk if you have any issues!
